types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,291 @@
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_axis event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.source = source,
+		.delta = wl_fixed_to_double(value),
+		.delta_discrete = discrete
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.axis, &event);
+}
+
+static void virtual_pointer_destroy_resource(struct wl_resource *resource) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	wlr_signal_emit_safe(&pointer->events.destroy, pointer);
+	wl_list_remove(&pointer->link);
+	wlr_input_device_destroy(&pointer->input_device);
+	free(pointer);
+}
+
+static void virtual_pointer_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zwlr_virtual_pointer_v1_interface virtual_pointer_impl = {
+	.motion = virtual_pointer_motion,
+	.motion_absolute = virtual_pointer_motion_absolute,
+	.button = virtual_pointer_button,
+	.axis = virtual_pointer_axis,
+	.frame = virtual_pointer_frame,
+	.axis_source = virtual_pointer_axis_source,
+	.axis_stop = virtual_pointer_axis_stop,
+	.axis_discrete = virtual_pointer_axis_discrete,
+	.destroy = virtual_pointer_destroy,
+};
+
+static const struct zwlr_virtual_pointer_manager_v1_interface manager_impl;
+
+static struct wlr_virtual_pointer_manager_v1 *manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_virtual_pointer_manager_v1_interface, &manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void virtual_pointer_manager_create_virtual_pointer(
+		struct wl_client *client, struct wl_resource *resource,
+		struct wl_resource *seat, uint32_t id) {
+	struct wlr_virtual_pointer_manager_v1 *manager = manager_from_resource(resource);
+
+	struct wlr_virtual_pointer_v1 *virtual_pointer = calloc(1,
+		sizeof(struct wlr_virtual_pointer_v1));
+	if (!virtual_pointer) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_pointer *pointer = calloc(1, sizeof(struct wlr_pointer));
+	if (!pointer) {
+		wlr_log(WLR_ERROR, "Cannot allocate wlr_pointer");
+		free(virtual_pointer);
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wlr_pointer_init(pointer, NULL);
+
+	struct wl_resource *pointer_resource = wl_resource_create(client,
+		&zwlr_virtual_pointer_v1_interface, wl_resource_get_version(resource),
+		id);
+	if (!pointer_resource) {
+		free(pointer);
+		free(virtual_pointer);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(pointer_resource, &virtual_pointer_impl,
+		virtual_pointer, virtual_pointer_destroy_resource);
+
+	wlr_input_device_init(&virtual_pointer->input_device,
+		WLR_INPUT_DEVICE_POINTER, &input_device_impl, "virtual pointer",
+		0x0, 0x0);
+
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_resource(seat);
```

Comment:
This function returns NULL if the seat is inert (destroyed by the compositor, but the client isn't aware of this yet). We'll probably want to create an inert virtual pointer in this case.